File: compiler.py
from lexer import *
import os
import shutil
from pathlib import Path
import sys
import subprocess
import gccdocksparser
from logs import *
from datetime import datetime
from headergenerator import *

try:
    absp = os.path.abspath(__file__).replace("compiler.py", "")

    floader_dir = sys.argv[1]

    try:
        os.makedirs(floader_dir + "/logs")
    except: pass

    log_path = floader_dir + "/logs/" + datetime.now().strftime("%Y-%m-%d_%H-%M-%S") + ".log"

    clear_log()

    def extract_additional_files(file_path):
            with open(file_path, 'r') as file:
                additional_files = []
                start_tag_found = False

                for line in file:
                    if start_tag_found:
                        if line.strip() == '[Additional Files end]':
                            break
                        additional_files.append(line.strip())
                    elif line.strip() == '[Additional Files start]':
                        start_tag_found = True

                return additional_files

    print_log("Compiler started")

    try:
        os.makedirs(floader_dir + "/bin")
        print_log("Created bin directory")
    except FileExistsError:
        print_log("Bin directory already exists")
        pass

    bin_dir = floader_dir + "/bin"

    buildfiles = ""

    settings = {"filename":"",}
    
    print_log("Reading files in floader")

    print(os.listdir(floader_dir))

    variables = {}
    classes = []
    includes = []
    namespaces = []

    for name in os.listdir(floader_dir):

        filename, fileext = os.path.splitext(name)

        compile = True

        if fileext == ".psc":
            fileext = ".cpp"
            
        elif fileext == ".ponycfg":
            compile = False

        else:
            compile = False

        if compile:

            print_log("Compiling file " + name)

            with open(f"{floader_dir}/{name}", "r", encoding="utf-8") as f:
                lines = [line for line in f]
                code = "";
                for line in lines:
                    code = code + line

                with open(f"{bin_dir}/{filename}{fileext}", "w", encoding="utf-8") as file:
                    file.write("")
                    file.close()

                with open(f"{bin_dir}/{filename}{fileext}", "a", encoding="utf-8") as output:
                    output.write('#include <iostream>' + '\n')
                    output.write(f'#include "{absp}lib/lib.h"' + '\n')
                    output.write(f'#include "{absp}lib\linked_list.h"' + '\n')
                    output.write(f'#define FILE_INFO {name}' + '\n')
                    #output.write('using namespace std;' + '\n')

                    print_log("Lexical analysis")
                    tokens = lexical_analyzer(code)

                    tokens = remove_comments(tokens)

                    print(tokens)

                    for i in get_libs(tokens):
                        if i not in includes:
                            includes.append(i)

                    

                    print_log("Variable analysis")
                    for i in find_variables(tokens):
                        variables[i] = find_variables(tokens)[i]
                    print(variables)
                    

                    for i in includes:
                        temptokens = lexical_analyzer(open(f"{absp}lib/{i.strip()}/{i.strip()}.h", "r").read())
                        temptokens = remove_comments(temptokens)
                        tempvariables = find_variables(temptokens)
                        for j in tempvariables:
                            variables[j] = tempvariables[j]

                    
                    
                    print_log("Class analysis")
                    classes_list, updated_tokens = find_classes(tokens)
                    for i in classes_list:
                        classes.append(i)

                    for i in includes:
                        tempclasses, temptokens = find_classes(remove_comments(lexical_analyzer(open(f"{absp}lib/{i.strip()}/{i.strip()}.h", "r").read())))
                        for j in tempclasses:
                            classes.append(j)

                    print_log("Namespace analysys")
                    for i in find_namespaces(tokens):
                        namespaces.append(i)
                    
                    for i in includes:
                        tempnamespaces = find_namespaces(remove_comments(lexical_analyzer(open(f"{absp}lib/{i.strip()}/{i.strip()}.h", "r").read())))
                        for j in tempnamespaces:
                            namespaces.append(j)

                    print_log("Translation")

                    tokens = translate_tokens(updated_tokens, variables, classes, namespaces, includes)

                    tokens = format_strings(tokens)
                    
                    print_log("Assembling code")
                    assembled_code = assemble_code(tokens)
                    print_log("Writing code to file")
                    output.write(assembled_code)
                    print_log("Finished compiling file " + name)

            generate_header(f"{bin_dir}/{filename}{fileext}", f"{bin_dir}/{filename}.h")
            buildfiles = buildfiles + f"{bin_dir}/{filename}{fileext} "
            buildfiles = buildfiles + f"{bin_dir}/{filename}.h "

    print_log("Reading configuration file")
    with open(f"{floader_dir}/.ponycfg") as cfg:
        additional_start = None
        additional_end = None
        for cfgline in cfg.readlines():
            if cfgline.split("=")[0].strip() == "[File Name]":
                settings["filename"] = cfgline.split("=")[1].strip()



    lib = "/lib" #Эта фигня нужна для дебага, gcc переодически теряет файлы в каталого lib. "/lib" - значение для каталога lib
    if lib != "":
        try:
            shutil.rmtree(f"{bin_dir}{lib}") #Удаляем каталог lib, если он существует
        except: pass
        try:
            os.mkdir(f"{bin_dir}{lib}")
        except FileExistsError:
            pass

    print_log("Copying built in libriaries")

    # buildfiles = buildfiles + f"{bin_dir}/lib/file_manager.cpp "

    buildfiles = buildfiles + f"{bin_dir}{lib}/run.cpp "

    
    # buildfiles = buildfiles + f"{bin_dir}/lib/math.cpp "

    print_log("Copying additional files")

    print_log("Copying additional built-in libriaries")
    for i in includes:

        shutil.copy(f"{absp}lib/{i.strip()}/{i.strip()}.cpp", bin_dir + lib)
        shutil.copy(f"{absp}lib/{i.strip()}/{i.strip()}.h", bin_dir + lib)
        buildfiles = buildfiles + f"{bin_dir}{lib}/{i}.cpp "
        buildfiles = buildfiles + f"{bin_dir}{lib}/{i}.h "

    print_log("Building executable")
    print_log("Writing run.h")

    with open(f"{bin_dir}{lib}/run.h", "w") as run:
        run.write(f'int magic(int argc, char *argv[]);')


    print_log("Writing run.cpp")

    with open(f"{bin_dir}{lib}/run.cpp", "w") as run:

        content = """"""

        content = content + f'#include "{absp}lib\linked_list.h"' + '\n'
        content = content + f'#include "{absp}lib\lib.h"' + '\n'
        content = content + f'#include "run.h"' + '\n'
        content = content + f'#include "{absp}lib\ponyexceptions.h"' + '\n' 

        content = content + """

            int main(int argc, char *argv[]) {
            try
            {
                return magic(argc, argv);

            }
            catch(const std::exception& e)
            {
                std::cerr << e.what();
            }
            catch (const PonyException& e) 
            {
                std::cerr << "Exception: " << e.what() << "File: " << e.getFile() << "Line: " << e.getLine() << std::endl;
            }


            }
            """
        run.write(content)
    
    # shutil.copy(f"{absp}lib/run.cpp", bin_dir+ lib)


    print_log("Starting gcc compyler")


    objectfiles = ""
    for i in buildfiles.split():

        if os.path.basename(i).split('.')[-1] == 'cpp':
            print_log("Compiling " + i)
            result = subprocess.run(f"{absp}ucrt64\\bin\\g++.exe -c  {i} -o {bin_dir}/{os.path.basename(i).split('.')[0]}.o -I{absp}lib -I{bin_dir}/lib", capture_output=True)
            objectfiles = objectfiles + f"{bin_dir}/{os.path.basename(i).split('.')[0]}.o "
            if result.returncode != 0:
                print(gccdocksparser.parse(result.stderr.decode()))
                raise Exception("Compilation failed!")


    print(f"{absp}ucrt64\\bin\\g++.exe -o {bin_dir}/{settings['filename']} {objectfiles} -L{absp}lib -I{absp}lib -lponylib -finput-charset=UTF-8 -lstdc++ ")
    compiler = subprocess.run(f"{absp}ucrt64\\bin\\g++.exe -o {bin_dir}/{settings['filename']} {objectfiles} -I{absp}lib -I{bin_dir}/lib -L{absp}lib  -lponylib -finput-charset=UTF-8 -lstdc++ ", capture_output=True)

    # for i in os.listdir(bin_dir):
    #     if i.split('.')[-1] == 'o':
    #         os.remove(bin_dir + "/" + i)

    if compiler.returncode == 0:
        print_log("Compilation successful!") 
    else:
        print_log(compiler.stderr.decode())
        raise Exception("Compilation failed!")

except Exception as e:
    print_log("\n")
    print_error(str(e))
    
exit()

File: compiler.cpp
#include <string>
#include <iostream>
#include <filesystem>
#include <vector>
#include <fstream>
#include <unistd.h>
#include <cstdlib>
#include <array>

#include "lexer.h"
#include "headergenerator.h"
#include "libparcer.h"
#include "utils.h"
#include "gccdockparcer.h"

namespace fs = std::filesystem;
using namespace std;

void remove_files_and_exit(int exit_code, string bin_directory){

    for (const auto & entry : fs::directory_iterator(bin_directory))
        {
            std::string file_path = entry.path().u8string();
        if (file_path.substr(file_path.find_last_of(".") + 1) == "o"){
            fs::remove(file_path);
        }
    }
    exit(exit_code);
}

std::pair<int, std::string> execute_command(const std::string& command) {
    FILE* pipe = _popen((command + " 2>&1").c_str(), "r");
    if (!pipe) {
        throw std::runtime_error("Ошибка при открытии pipe.");
    }
    char buffer[128];
    std::string result;
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        result += buffer;
    }
    int exit_code = _pclose(pipe);
    return make_pair(exit_code, result);
}


int main(int argc, char *argv[])
{
    std::filesystem::path exePath = std::filesystem::absolute(std::filesystem::path(argv[0]));

    std::filesystem::path exeDir = exePath.parent_path();

    filesystem::path directoryPath = exeDir;
    string stringpath = directoryPath.generic_string(); 

    std::vector<std::string> filestocompile;
    std::string dir_path = argv[1];
    std::vector<std::string> buildfiles;
    std::vector<std::string> included_libs;

    vector<string> secondary_build_files;

    string exit_file_name;

    vector<string> libs = get_libs(stringpath + "/lib");

    build_libs(stringpath + "/lib");

    ifstream cfg_file(stringpath + "/.ponycfg");
    string cfg_line ="";

    for (int i = 1; i < argc; i++) {
        std::cout << argv[i] << std::endl;
        if (strcmp(argv[i], "-o") == 0 && i + 1 < argc) {
            exit_file_name = argv[i + 1];
            i++;
        }
    }

    for (const auto & entry : fs::directory_iterator(dir_path))
        {
            std::string file_path = entry.path().u8string();
        if (file_path.substr(file_path.find_last_of(".") + 1) == "psc"){
            filestocompile.push_back(file_path);
        }
    }

    fs::remove_all(dir_path + "/bin");
    if(!fs::is_directory(dir_path + "/bin"))
    {
        fs::create_directory(dir_path + "/bin");
        fs::create_directory(dir_path + "/bin/lib");
    }
        
    for (std::string file : filestocompile){

        std::ifstream file_stream(file);

        std::string file_content;
        std::string line;
        while(std::getline(file_stream, line)){
            file_content += line + "\n";
        }
        line.clear();
        file_stream.close();

        std::vector<std::string> tokens = lexical_analyzer(file_content);
        file_content.clear();

        file = file.substr(file.find_last_of("/"));

        std::string file_name = fs::path(file).stem().u8string();

        std::vector<std::string> includes =  find_includes(tokens);

        for(string include : includes){
            if(is_vector_contain_string(libs, include) && !is_vector_contain_string(included_libs, include))
                included_libs.push_back(include);
        }

        vector<string> classes = find_classes(tokens);

        vector<string> functions = find_functions(tokens);

        vector<string> variables = find_variables(tokens, classes);

        format_strings(tokens);

        vector<string> namespaces = find_namespaces(tokens);

        translate_tokens(tokens, variables, classes, namespaces, includes);

        std::string output_file_content = "";

        output_file_content += "#include \"lib.h\" \n";

        output_file_content += "#include \"linked_list.h\" \n";

        output_file_content += "#define FILE_INFO " + file_name + ".psc \n";

        std::string output_file_path =  dir_path + "/bin/" + file_name + ".cpp";

        ofstream output_file(output_file_path);

        if(output_file.is_open())
        {

            output_file_content += assemble_tokens(tokens);

            output_file << output_file_content;

            output_file.close();

            tokens.clear();

            std::string header_file_path = dir_path + "/bin/" + file_name + ".h";

            generate_cpp_header(output_file_path, header_file_path);

            fprintf(stderr, "Compiled %s\n", file.c_str());
        }else{
            cout << "Error opening file!" << endl;
        }

        buildfiles.push_back(dir_path + "/bin/" + file_name + ".cpp");

        output_file.clear();
        output_file_content.clear();
        file_stream.clear();

    }

    fs::copy(stringpath + "/lib/run.cpp", dir_path + "/bin/lib/run.cpp");
    fs::copy(stringpath + "/lib/run.h", dir_path + "/bin/lib/run.h");

    secondary_build_files.push_back(dir_path + "/bin/lib/run.cpp");

    string libs_path_arguments = "";

    for(string lib : included_libs){
        libs_path_arguments += "-I" + stringpath + "/lib/" + lib + " ";
        secondary_build_files.push_back(stringpath + "/lib/" + lib + "/" + lib +".cpp");
    }

    string build_arguments = "-L" + stringpath +"/lib -I" + stringpath+ "/lib " + libs_path_arguments + " -lponylib -finput-charset=UTF-8 -lstdc++";

    for(string file : buildfiles){
        std::string file_name = fs::path(file).stem().u8string();

        cout << "Compiling file " + file << endl;
        pair<int, string> builder_output = execute_command(stringpath + "/ucrt64/bin/g++.exe -c " + file + " -o " + dir_path + "/bin/" + file_name + ".o " + build_arguments);

        string output = builder_output.second;
        int exit_code = builder_output.first;

        cout << parse(output) << endl;

        if(exit_code != 0){
            cout << "Error compiling file!" << endl;
            remove_files_and_exit(1, dir_path + "/bin");
        }

        secondary_build_files.push_back(dir_path + "/bin/" + file_name + ".o");
    }
    
    string secondary_build_files_string;
    for(string file : secondary_build_files){
        secondary_build_files_string += file + " ";
    }

    cout << "Executable file name " + exit_file_name << endl;
    cout << "Building executable..." << endl;

    string secondray_build_command = stringpath + "/ucrt64/bin/g++ " + secondary_build_files_string + " -o " + dir_path + "/bin/"  + exit_file_name + " " + build_arguments;

    pair<int, string> output_pair = execute_command(secondray_build_command);

    string output = output_pair.second;
    int exit_code = output_pair.first;

    cout << parse(output)<< endl;


    if(exit_code != 0){
        cout << "Error building executable!" << endl;
        remove_files_and_exit(1, dir_path + "/bin");
    }
    

    cout << "Executable built successfully!" << endl;

    remove_files_and_exit(0, dir_path + "/bin");

    return 0;

}
